File: src/main/java/com/micronaut/bug/service/EnumConverterConfig.java
package com.micronaut.bug.service;

import com.micronaut.bug.api.EnumParam;
import io.micronaut.context.annotation.Bean;
import io.micronaut.context.annotation.Factory;
import io.micronaut.core.convert.TypeConverter;

import java.util.Optional;

@Factory
public class EnumConverterConfig {

    @Bean
    public TypeConverter<String, EnumParam> strToEnumConverter() {
        return (v, c, ctx) -> Optional.of(EnumParam.fromValue(Integer.valueOf(v)));
    }

    @Bean
    public TypeConverter<EnumParam, String> enumToStrConverter() {
        return (v, c, ctx) -> Optional.of(v.toString());
    }
}

File: src/main/kotlin/com/micronaut/bug/controller/EnumConverterConfig.kt
package com.micronaut.bug.controller

import io.micronaut.context.annotation.Bean
import io.micronaut.context.annotation.Factory
import io.micronaut.core.convert.TypeConverter
import java.util.Optional

@Factory
class EnumConverterConfig {

    @Bean
    fun strToEnumConverter() = TypeConverter<String?, EnumParam?> { v, _, _ -> Optional.of(EnumParam.fromValue(v.toInt())) }

    @Bean
    fun enumToStrConverter() = TypeConverter<EnumParam?, String?> { v, _, _ -> Optional.of(v.value.toString()) }
}
